java-checks/src/main/java/org/sonar/java/checks/DeadStoreCheck.java

```diff
@@ -0,0 +1,185 @@
+    return ImmutableList.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodTree methodTree = (MethodTree) tree;
+    if (methodTree.block() == null) {
+      return;
+    }
+
+    //Exclude try statements with finally as CFG is incorrect for those and lead to false positive
+    if(hasTryFinallyWithLocalVar(methodTree.block(), methodTree.symbol())) {
+      return;
+    }
+
+    Symbol.MethodSymbol methodSymbol = methodTree.symbol();
+    CFG cfg = CFG.build(methodTree);
+    LiveVariables liveVariables = LiveVariables.analyze(cfg);
+    for (CFG.Block block : cfg.blocks()) {
+      checkElements(Lists.reverse(block.elements()), liveVariables.getOut(block), methodSymbol);
+    }
+  }
+
+  private void checkElements(List<Tree> elements, Set<Symbol> blockOut, Symbol.MethodSymbol methodSymbol) {
+    Set<Symbol> out = new HashSet<>(blockOut);
+    Set<Tree> assignmentLHS = new HashSet<>();
+    for (Tree element : elements) {
+      Symbol symbol;
+      switch (element.kind()) {
+        case ASSIGNMENT:
+          ExpressionTree lhs = ((AssignmentExpressionTree) element).variable();
+          if (lhs.is(Tree.Kind.IDENTIFIER)) {
+            symbol = ((IdentifierTree) lhs).symbol();
+            if (isLocalVariable(symbol) && !out.contains(symbol)) {
+              addIssue(element, "Remove this useless assignment to local variable \"" + symbol.name() + "\".");
+            }
+            assignmentLHS.add(lhs);
+            out.remove(symbol);
+          }
+          break;
+        case IDENTIFIER:
+          symbol = ((IdentifierTree) element).symbol();
+          if (!assignmentLHS.contains(element) && isLocalVariable(symbol)) {
+            out.add(symbol);
+          }
+          break;
+        case VARIABLE:
+          VariableTree localVar = (VariableTree) element;
+          symbol = localVar.symbol();
+          if(localVar.initializer() != null && !out.contains(symbol)) {
+            addIssue(element, "Remove this useless assignment to local variable \"" + symbol.name() + "\".");
+          }
+          out.remove(symbol);
+          break;
+        case NEW_CLASS:
+          ClassTree body = ((NewClassTree) element).classBody();
+          if(body != null) {
+            List<Symbol> usedLocalVar = getUsedLocalVarInSubTree(body, methodSymbol);
+            out.addAll(usedLocalVar);
+          }
+          break;
+        case LAMBDA_EXPRESSION:
+          LambdaExpressionTree lambda = (LambdaExpressionTree) element;
+          out.addAll(getUsedLocalVarInSubTree(lambda.body(), methodSymbol));
+          break;
+        case TRY_STATEMENT:
+          TryStatementTree tryStatement = (TryStatementTree) element;
+          for (CatchTree catchTree : tryStatement.catches()) {
+            out.addAll(getUsedLocalVarInSubTree(catchTree, methodSymbol));
+          }
+          break;
+        case CLASS:
+        case ENUM:
+        case ANNOTATION_TYPE:
+        case INTERFACE:
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) End this switch case with an unconditional break, return or throw statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS128)